Create section for optimistic block (#13500)
near/nearcore: docs/architecture/how/optimistic_block.md

# Optimistic block

This short document will tell you about what optimistic block is and why it was introduced with protocol version 77.

## Improving Stateless Validation Latency with Optimistic Block

### Why 2× Latency Existed Before v77 — and How the Idea of Optimistic Block Emerged

![image](https://raw.githubusercontent.com/near/nearcore/7f0299041e3c3c9241a8a300900ec9a8d07fd7c1/docs/images/optimistic_block_before.png)

In this diagram: yellow = data; rectangle = process. Arrow = dependency on a process; dashed arrow = production and dependency on data.

> This is the critical path today between two consecutive blocks. The two red blocks are the slow parts. It's better to have only one of them on the critical path.

Reviewer: ```suggestion
This was the critical path between two consecutive blocks. The two red blocks are the slow parts. It's better to have only one of them on the critical path.
```

    * A chunk does not contain the incoming receipts of that shard; rather it contains the outgoing receipts of the previous chunk of that shard. In order to collect the incoming receipts, other chunks in the same block are needed. During optimistic execution, these are obtained from the other chunks that are received via ShardsManager.
  * The optimistic chunk application should produce the exact same result as the real chunk application as long as the optimistic block matches all of the real block’s fields, and the chunks used for optimistic application are exactly the chunks included in the real block (in other words, no chunk producer double-signed, and no chunks were missed).
* **Fallback and safety:** It is possible for the optimistic chunk application to mispredict the actual block produced. Therefore, the original chunk application logic still needs to be executed in that case.
  * To implement cleanly, the optimistic chunk application shall serve only as a cache for the actual chunk application. It should not write any data or do anything else other than acting as a cache. The actual chunk application always happens except it may be skipped if the inputs match exactly.

> * **Changes to block timestamp:** Note that the way validators choose the block timestamp must be changed, as it needs to be predicted by the time the block producer produces the optimistic block.
>   * The timestamp is not a critical field; it only needs to be reasonable (tethered to real world time) and monotonically increasing.
>   * In the current implementation, the current timestamp is used when producing the optimistic block.
>   * This is not a protocol change as the block producer has freedom in picking this value.

Reviewer: ```suggestion
* **Changes to block timestamp:** Note that the way validators choose the block timestamp must be changed, as it needs to be predicted by the time the block producer produces the optimistic block.
  * The timestamp is not a critical field; it only needs to be reasonable (tethered to real world time) and monotonically increasing.
  * In the current implementation, the current timestamp is used when producing the optimistic block.
  * This is not a protocol change as the block producer has freedom in picking this value.
  * If optimistic block was produced, the block and optimitstic block timestamp must be the same. Otherwise, the optimistic chunk application will be invalidated. 
```

Reviewer: Here I just added the last line
